File: apps/api/app/api/schemas.py
"""Request and response schemas for /api/v1.

Money crosses the wire as a decimal *string*, never a JSON number: a JSON number
is a float in most clients, and 982.45 does not survive that round trip.
"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_serializer

__all__ = [
    "ApproveMatchRequest",
    "CloseRunRequest",
    "CreateReconciliationRequest",
    "EvidenceResponse",
    "ExceptionResponse",
    "ManualMatchRequest",
    "MappingRequest",
    "MatchResponse",
    "Problem",
    "ReconciliationResponse",
    "RejectMatchRequest",
    "ReopenRunRequest",
    "RunResponse",
    "RunSummaryResponse",
    "StartRunRequest",
    "TransactionResponse",
    "TransitionExceptionRequest",
]


class Base(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class Problem(Base):
    """RFC 7807-shaped error body."""

    type: str = "about:blank"
    title: str
    status: int
    detail: str
    code: str = ""
    instance: str | None = None


class MoneyMixin(BaseModel):
    @field_serializer("*", when_used="json")
    def _serialise_decimal(self, value: Any) -> Any:
        return str(value) if isinstance(value, Decimal) else value


# ---------------------------------------------------------------------------
# Files and mapping
# ---------------------------------------------------------------------------


class ColumnMappingRequest(Base):
    source_column: str
    canonical_field: str
    date_format: str | None = None
    number_format: str | None = None
    negate: bool = False


class MappingRequest(Base):
    source_system: str
    columns: list[ColumnMappingRequest]
    static_values: dict[str, str] = Field(default_factory=dict)
    debit_column: str | None = None
    credit_column: str | None = None


class FileResponse(Base):
    id: UUID
    filename: str
    byte_size: int
    checksum: str
    status: str
    row_count: int | None
    encoding: str | None
    delimiter: str | None
    sheet_name: str | None
    created_at: datetime


class IngestResponse(Base):
    file_id: UUID
    transactions_created: int
    duplicates_skipped: int
    rows_failed: int
    quality_level: str
    quality_blocking: bool
    findings: list[dict[str, Any]]


# ---------------------------------------------------------------------------
# Transactions
# ---------------------------------------------------------------------------


class TransactionResponse(MoneyMixin, Base):
    id: UUID
    source_system: str
    source_record_id: str
    transaction_date: date | None
    posting_date: date | None
    value_date: date | None
    amount: Decimal
    currency: str
    debit_credit: str
    description: str | None
    reference: str | None
    counterparty_name: str | None
    invoice_number: str | None
    settlement_id: str | None
    external_transaction_id: str | None
    gross_amount: Decimal | None
    fee_amount: Decimal | None
    net_amount: Decimal | None
    status: str | None
    imported_at: datetime


class LineageResponse(Base):
    transaction_id: UUID
    source_record_id: str
    source_checksum: str
    records: list[dict[str, str]]


# ---------------------------------------------------------------------------
# Reconciliations and runs
# ---------------------------------------------------------------------------


class CreateReconciliationRequest(Base):
    slug: str = Field(min_length=2, max_length=64, pattern=r"^[a-z0-9][a-z0-9-]*$")
    name: str = Field(min_length=1, max_length=255)
    template: str | None = None
    config: dict[str, Any] | None = None
    side_a_connection_id: UUID | None = None
    side_b_connection_id: UUID | None = None
    entity: str = "default"


class ReconciliationResponse(Base):
    id: UUID
    slug: str
    name: str
    template: str | None
    config: dict[str, Any]
    config_version: str
    config_hash: str
    entity: str
    version: int
    created_at: datetime


class StartRunRequest(Base):
    period_start: date | None = None
    period_end: date | None = None


class RunResponse(Base):
    id: UUID
    reconciliation_id: UUID
    status: str
    period_start: date | None
    period_end: date | None
    started_at: datetime | None
    finished_at: datetime | None
    result_hash: str | None
    version: int
    replayed: bool = False


class RunSummaryResponse(MoneyMixin, Base):
    run_id: UUID
    reconciliation_id: UUID
    status: str
    period_start: date | None
    period_end: date | None
    currency: str
    side_a_balance: Decimal
    side_b_balance: Decimal
    difference: Decimal
    matched_amount: Decimal
    matched_transaction_count: int
    auto_matched_count: int
    human_approved_count: int
    suggested_count: int
    exception_count: int
    high_risk_exception_count: int
    unmatched_a_count: int
    unmatched_b_count: int
    oldest_exception_age_days: int | None
    completion_pct: float
    auto_match_rate: float
    # Never shown without the false-match rate beside it (spec section 48).
    false_match_rate: float | None
    manual_review_rate: float


# ---------------------------------------------------------------------------
# Matches
# ---------------------------------------------------------------------------


class MatchMemberResponse(MoneyMixin, Base):
    transaction_id: UUID
    side: str
    allocated_amount: Decimal


class MatchReasonResponse(Base):
    code: str
    contribution: float
    description: str


class MatchResponse(MoneyMixin, Base):
    id: UUID
    run_id: UUID
    cardinality: str
    status: str
    decision: str
    confidence: float
    score: float
    currency: str
    total_amount: Decimal
    rule_id: str | None
    rule_version: str | None
    rule_set_version: str | None
    model_version: str | None
    engine_stage: str
    competing_candidate_count: int
    members: list[MatchMemberResponse]
    reasons: list[MatchReasonResponse]
    warnings: list[dict[str, str]]
    approved_by: UUID | None
    approved_at: datetime | None
    override_reason: str | None
    version: int
    explanation: str = ""


class ApproveMatchRequest(Base):
    expected_version: int = Field(ge=1)
    reason: str | None = None


class RejectMatchRequest(Base):
    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=3, max_length=2000)


class UnmatchRequest(Base):
    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=3, max_length=2000)


class ManualMatchRequest(Base):
    run_id: UUID
    side_a_ids: list[UUID] = Field(min_length=1)
    side_b_ids: list[UUID] = Field(min_length=1)
    reason: str = Field(min_length=3, max_length=2000)


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class ExceptionResponse(MoneyMixin, Base):
    id: UUID
    reconciliation_run_id: UUID
    transaction_ids: list[UUID]
    category: str
    severity: str
    status: str
    amount_exposure: Decimal | None
    currency: str | None
    title: str
    detail: str
    reason_codes: list[str]
    owner_user_id: UUID | None
    first_detected_at: datetime
    due_at: datetime | None
    proposed_resolution: str | None
    resolution_code: str | None
    proposed_by_actor_type: str | None
    requires_approval: bool
    closed_by: UUID | None
    closed_at: datetime | None
    version: int


class TransitionExceptionRequest(Base):
    status: Literal[
        "TRIAGED",
        "ASSIGNED",
        "INVESTIGATING",
        "PROPOSED_RESOLUTION",
        "AWAITING_APPROVAL",
        "RESOLVED",
        "CLOSED",
        "BLOCKED",
        "ESCALATED",
        "REOPENED",
    ]
    expected_version: int | None = None
    reason: str | None = None
    owner_user_id: UUID | None = None
    resolution_code: str | None = None
    proposed_resolution: str | None = None


class AssignExceptionRequest(Base):
    owner_user_id: UUID
    expected_version: int | None = None


class CommentRequest(Base):
    body: str = Field(min_length=1, max_length=8000)


class CommentResponse(Base):
    id: UUID
    exception_id: UUID
    author_id: UUID
    body: str
    created_at: datetime


class EvidenceResponse(Base):
    id: UUID
    exception_id: UUID | None
    filename: str
    mime_type: str
    byte_size: int
    sha256: str
    kind: str
    uploaded_by: UUID
    uploaded_at: datetime
    download_url: str | None = None


class ProposeResolutionRequest(Base):
    proposed_resolution: str = Field(min_length=3, max_length=4000)
    resolution_code: str | None = None
    expected_version: int | None = None


# ---------------------------------------------------------------------------
# Close
# ---------------------------------------------------------------------------


class CloseRunRequest(Base):
    expected_version: int = Field(ge=1)


class ReopenRunRequest(Base):
    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=3, max_length=2000)


class CloseCertificateResponse(MoneyMixin, Base):
    run_id: UUID
    period: str
    status: str
    book_balance: Decimal
    external_balance: Decimal
    difference: Decimal
    unresolved_items: int
    closed_by: UUID
    closed_at: datetime
    config_version: str
    snapshot_hash: str
    certificate_hash: str


class InvestigationRequest(Base):
    question: str = Field(min_length=3, max_length=500)

File: apps/web/lib/schemas.ts
import { z } from "zod";

const uuid = z.string().uuid();
const isoDate = z.string().nullable();
const money = z.string().regex(/^-?\d+(?:\.\d+)?$/, "Money must be a decimal string");

export const meSchema = z.object({
  user_id: uuid.nullable(),
  email: z.string().nullable(),
  tenant_id: uuid,
  actor_type: z.string(),
  roles: z.array(z.string()),
  permissions: z.array(z.string()),
  approval_limit: z.number().nullable(),
  ai: z.object({ enabled: z.boolean(), policy: z.string(), provider: z.string() }),
});

export const reconciliationSchema = z.object({
  id: uuid,
  slug: z.string(),
  name: z.string(),
  template: z.string().nullable(),
  config: z.record(z.string(), z.unknown()),
  config_version: z.string(),
  config_hash: z.string(),
  entity: z.string(),
  version: z.number().int(),
  created_at: z.string(),
});

export const runSchema = z.object({
  id: uuid,
  reconciliation_id: uuid,
  status: z.string(),
  period_start: isoDate,
  period_end: isoDate,
  started_at: z.string().nullable(),
  finished_at: z.string().nullable(),
  result_hash: z.string().nullable(),
  version: z.number().int(),
  replayed: z.boolean().default(false),
});

export const runSummarySchema = z.object({
  run_id: uuid,
  reconciliation_id: uuid,
  status: z.string(),
  period_start: isoDate,
  period_end: isoDate,
  currency: z.string(),
  side_a_balance: money,
  side_b_balance: money,
  difference: money,
  matched_amount: money,
  matched_transaction_count: z.number().int(),
  auto_matched_count: z.number().int(),
  human_approved_count: z.number().int(),
  suggested_count: z.number().int(),
  exception_count: z.number().int(),
  high_risk_exception_count: z.number().int(),
  unmatched_a_count: z.number().int(),
  unmatched_b_count: z.number().int(),
  oldest_exception_age_days: z.number().int().nullable(),
  completion_pct: z.number(),
  auto_match_rate: z.number(),
  false_match_rate: z.number().nullable(),
  manual_review_rate: z.number(),
});

export const transactionSchema = z.object({
  id: uuid,
  source_system: z.string(),
  source_record_id: z.string(),
  transaction_date: isoDate,
  posting_date: isoDate,
  value_date: isoDate,
  amount: money,
  currency: z.string(),
  debit_credit: z.string(),
  description: z.string().nullable(),
  reference: z.string().nullable(),
  counterparty_name: z.string().nullable(),
  invoice_number: z.string().nullable(),
  settlement_id: z.string().nullable(),
  external_transaction_id: z.string().nullable(),
  gross_amount: money.nullable(),
  fee_amount: money.nullable(),
  net_amount: money.nullable(),
  status: z.string().nullable(),
  imported_at: z.string(),
});

export const matchSchema = z.object({
  id: uuid,
  run_id: uuid,
  cardinality: z.string(),
  status: z.string(),
  decision: z.string(),
  confidence: z.number(),
  score: z.number(),
  currency: z.string(),
  total_amount: money,
  rule_id: z.string().nullable(),
  rule_version: z.string().nullable(),
  rule_set_version: z.string().nullable(),
  model_version: z.string().nullable(),
  engine_stage: z.string(),
  competing_candidate_count: z.number().int(),
  members: z.array(z.object({ transaction_id: uuid, side: z.string(), allocated_amount: money })),
  reasons: z.array(z.object({ code: z.string(), contribution: z.number(), description: z.string() })),
  warnings: z.array(z.record(z.string(), z.string())),
  approved_by: uuid.nullable(),
  approved_at: z.string().nullable(),
  override_reason: z.string().nullable(),
  version: z.number().int(),
  explanation: z.string(),
});

export const exceptionSchema = z.object({
  id: uuid,
  reconciliation_run_id: uuid,
  transaction_ids: z.array(uuid),
  category: z.string(),
  severity: z.string(),
  status: z.string(),
  amount_exposure: money.nullable(),
  currency: z.string().nullable(),
  title: z.string(),
  detail: z.string(),
  reason_codes: z.array(z.string()),
  owner_user_id: uuid.nullable(),
  first_detected_at: z.string(),
  due_at: z.string().nullable(),
  proposed_resolution: z.string().nullable(),
  resolution_code: z.string().nullable(),
  proposed_by_actor_type: z.string().nullable(),
  requires_approval: z.boolean(),
  closed_by: uuid.nullable(),
  closed_at: z.string().nullable(),
  version: z.number().int(),
});

export const fileSchema = z.object({
  id: uuid,
  filename: z.string(),
  byte_size: z.number().int(),
  checksum: z.string(),
  status: z.string(),
  row_count: z.number().int().nullable(),
  encoding: z.string().nullable(),
  delimiter: z.string().nullable(),
  sheet_name: z.string().nullable(),
  created_at: z.string(),
});

export const lineageSchema = z.object({
  transaction_id: uuid,
  source_record_id: z.string(),
  source_checksum: z.string(),
  records: z.array(z.record(z.string(), z.string())),
});

export const commentSchema = z.object({
  id: uuid,
  exception_id: uuid,
  author_id: uuid,
  body: z.string(),
  created_at: z.string(),
});

export const evidenceSchema = z.object({
  id: uuid,
  exception_id: uuid.nullable(),
  filename: z.string(),
  mime_type: z.string(),
  byte_size: z.number().int(),
  sha256: z.string(),
  kind: z.string(),
  uploaded_by: uuid,
  uploaded_at: z.string(),
  download_url: z.string().nullable(),
});

export type Me = z.infer<typeof meSchema>;
export type Reconciliation = z.infer<typeof reconciliationSchema>;
export type Run = z.infer<typeof runSchema>;
export type RunSummary = z.infer<typeof runSummarySchema>;
export type Transaction = z.infer<typeof transactionSchema>;
export type Match = z.infer<typeof matchSchema>;
export type ReconException = z.infer<typeof exceptionSchema>;
export type SourceFile = z.infer<typeof fileSchema>;
export type Lineage = z.infer<typeof lineageSchema>;
export type Comment = z.infer<typeof commentSchema>;
export type Evidence = z.infer<typeof evidenceSchema>;
